### PR TITLE
Mps adaptive

### PR DESCRIPTION
**Pull request by AdriaPerezCulubret**
Added compatibility with latest jobqueues update to use nividia_mps with Slurm in Adaptive sampling.

### CODE REVIEW

File: htmd/adaptive/adaptive.py [hunk 1]
```diff
@@ -154,7 +161,25 @@ def run(self):
...
+                    if self.mps > 0:
+                        from jobqueues.slurmqueue import SlurmQueue
+
+                        if not isinstance(self.app, SlurmQueue):
+                            raise ValueError(
+                                "Multiple processes per simulation only supported with SlurmQueue."
+                            )
+                        input_dirs = natsorted(glob(path.join(self.inputpath, "e1s*")))
+                        for i in range(0, len(input_dirs), self.mps):
+                            if len(input_dirs[i : i + self.mps]) > 1:
```
Comment (stefdoerr) [inline htmd/adaptive/adaptive.py:173]: I'm rather sure you will get an index error here if your simulations are not divisible by self.mps
Use `min(i+self.mps, len(input_dirs))` to get the max of the range

File: htmd/adaptive/adaptive.py [hunk 2]
```diff
@@ -154,7 +161,25 @@ def run(self):
...
                 if not self.dryrun:
-                    self.app.submit(natsorted(glob(path.join(self.inputpath, "e1s*"))))
+                    if self.mps > 0:
+                        from jobqueues.slurmqueue import SlurmQueue
+
+                        if not isinstance(self.app, SlurmQueue):
+                            raise ValueError(
+                                "Multiple processes per simulation only supported with SlurmQueue."
+                            )
+                        input_dirs = natsorted(glob(path.join(self.inputpath, "e1s*")))
```
Comment (stefdoerr) [inline htmd/adaptive/adaptive.py:171]: move this out of the two if/else conditions. It's identical

File: htmd/adaptive/adaptive.py (hunk 1 above)
Comment (stefdoerr) [reply htmd/adaptive/adaptive.py:173]:
```py
batch = input_dirs[i:min(i+self.mps, len(input_dirs))]
self.app.submit(batch, nvidia_mps=len(batch) > 1)
```
Then you skip that else-if as well

**Comment by AdriaPerezCulubret** [comment]
Forgot there's two submits, so now I moved the mps part into a function, it's a bit cleaner now

File: htmd/adaptive/adaptive.py [hunk 3]
```diff
@@ -203,7 +214,10 @@ def run(self):
                             path.join(self.inputpath, "e" + str(epoch + 1) + "s*")
                         )
                         try:
-                            self.app.submit(natsorted(newsims))
+                            if self.mps > 0:
+                                self._submit_mps(newsims)
+                            else:
+                                self.app.submit(natsorted(newsims))
```
Comment (stefdoerr) [inline htmd/adaptive/adaptive.py:220]: just move the if/else inside of the _submit_mps (rename it to just _submit). Cleans up the `run` function.

```
    def _submit(self, newsims):
        from jobqueues.slurmqueue import SlurmQueue
        
        if self.mps == 0:
            self.app.submit(natsorted(newsims))
            return

        if not isinstance(self.app, SlurmQueue):
            raise ValueError(
                "Multiple processes per simulation only supported with SlurmQueue."
            )
        for i in range(0, len(newsims), self.mps):
            batch = newsims[i : min(i + self.mps, len(newsims))]
            self.app.submit(batch, nvidia_mps=len(batch) > 1)
```

Comment (AdriaPerezCulubret) [reply htmd/adaptive/adaptive.py]: Done!

File: htmd/adaptive/adaptive.py [hunk 4]
```diff
@@ -153,8 +160,9 @@ def run(self):
             if epoch == 0 and self.generatorspath:
                 logger.info("Epoch 0, generating first batch")
                 self._init()
+                input_dirs = natsorted(glob(path.join(self.inputpath, "e1s*")))
                 if not self.dryrun:
-                    self.app.submit(natsorted(glob(path.join(self.inputpath, "e1s*"))))
+                    self.app.submit(input_dirs)
```
Comment (stefdoerr) [inline htmd/adaptive/adaptive.py:165]: you forgot to use the _submit method in the two places though :D

**Review by stefdoerr — APPROVED** [review]

**Pull request merged**